enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+        } else {
+          op->emitWarning(
+              "annotating icv for op that produces multiple results");
+          icv = false;
+        }
+        op->setAttr("enzyme.icv", BoolAttr::get(ctx, icv));
+      });
+    });
+  }
+  callee.walk([&](Operation *op) {
+    if (op->hasAttr("tag")) {
+      errs() << "  " << op->getAttr("tag") << ": ";
+      for (OpResult opResult : op->getResults()) {
+        errs() << (isConstantValue(opResult) ? "Constant" : "Active") << "\n";
+      }
+    }
+    if (verbose) {
+      // Annotate each op's results with its value activity states
+      for (OpResult result : op->getResults()) {
+        auto forwardValueActivity =
+            solver.lookupState<ForwardValueActivity>(result);
+        if (forwardValueActivity) {
+          std::string dest, key{"fva"};
+          llvm::raw_string_ostream os(dest);
+          if (op->getNumResults() != 1)
+            key += result.getResultNumber();
+          forwardValueActivity->getValue().print(os);
+          op->setAttr(key, StringAttr::get(op->getContext(), dest));
+        }
+
+        auto backwardValueActivity =
+            solver.lookupState<BackwardValueActivity>(result);
+        if (backwardValueActivity) {
+          std::string dest, key{"bva"};
+          llvm::raw_string_ostream os(dest);
+          if (op->getNumResults() != 1)
+            key += result.getResultNumber();
+          backwardValueActivity->getValue().print(os);
+          op->setAttr(key, StringAttr::get(op->getContext(), dest));
+        }
+      }
+    }
+  });
+
+  if (verbose) {
+    // Annotate function attributes
+    for (BlockArgument arg : callee.getArguments()) {
+      auto backwardValueActivity =
+          solver.lookupState<BackwardValueActivity>(arg);
+      if (backwardValueActivity) {
+        std::string dest;
+        llvm::raw_string_ostream os(dest);
+        backwardValueActivity->getValue().print(os);
+        callee.setArgAttr(arg.getArgNumber(), "enzyme.bva",
+                          StringAttr::get(callee->getContext(), dest));
+      }
+    }
+
+    for (Operation *returnOp : returnOps) {
+      auto *state = solver.lookupState<ForwardMemoryActivity>(returnOp);
+      if (state)
+        errs() << "forward end state:\n" << *state << "\n";
+      else
+        errs() << "state was null\n";
+    }
+
+    auto startState = solver.lookupState<BackwardMemoryActivity>(
+        &callee.getFunctionBody().front().front());
+    if (startState)
+      errs() << "backwards end state:\n" << *startState << "\n";
+    else
+      errs() << "backwards end state was null\n";
+  }
+}
+
+void enzyme::runDataFlowActivityAnalysis(
+    FunctionOpInterface callee, ArrayRef<enzyme::Activity> argumentActivity,
+    bool print, bool verbose, bool annotate) {
+  SymbolTableCollection symbolTable;
+  DataFlowSolver solver;
+
+  solver.load<enzyme::PointsToPointerAnalysis>();
+  solver.load<enzyme::AliasAnalysis>(callee.getContext());
```

Comment:
(*) do we actually need these while computing the activity analyses, or can we finish them entirely beforehand? If so, we could abstract away the queries and compare the time it takes to run it in two stages vs. one stage.

Reply:
Activity analysis never feeds into alias analysis, so we can finish them entirely beforehand.

> we could abstract away the queries and compare the time it takes to run it in two stages vs. one stage.

Completely agree!